docs/rfcs/0006-top-sites-feature.md

```diff
@@ -0,0 +1,125 @@
+---
+layout: page
+title: Introduce a Top Sites Feature
+permalink: /rfc/0006-top-sites-feature
+---
+
+* Start date: 2020-07-30
+* RFC PR: [#7931](https://github.com/mozilla-mobile/android-components/pull/7931)
+
+## Summary
+
+Create a TopSitesFeature that abstracts out the co-ordination between the multiple data storages.
+
+## Motivation
+
+A large part of the top sites implementation is spread across A-C and Fenix. There is a lot of co-ordination between the History API and the current Top Sites API to manage the final state of what sites should be displayed that can exist as a single implementation.
+
+## Guide-level explanation
+
+To begin, we should rename the `TopSitesStorage`  to `PinnedSitesStorage` and create a `TopSitesFeature` that would orchestrate the different storages and present them to the app.
+
+We can then follow the architecture of existing features by using the presenter/interactor/view model, and introduce a `TopSitesStorage` which abstract out the complexity of the multiple data sources from `PinnedSitesStorage` and `PlacesHistoryStorage`.
+
+To allow the ability for adding a top site from different parts of the app (e.g. context menus), we introduce the `TopSitesUseCases` that acts on the storage directly. If the `TopSitesFeature` is started, it will then be notified by the `Observer` to update the UI.
+
+```kotlin
+
+/**
+ * Implemented by the application for displaying onto the UI.
+ */
+interface TopSitesView {
+  /**
+   * Update the UI.
+   */
+  fun displaySites(sites: List<TopSite>)
+
+  interface Listener {
+    /**
+     * Invoked by the UI for us to add to our storage.
+     */
+    fun onAddTopSite(topSite: TopSite)
+
+    /**
+     * Invoked by the UI for us to remove from our storage.
+     */
+    fun onRemoveTopSite(topSite: TopSite)
+  }
+}
+
+/**
+ * Abstraction layer above the multiple storages.
+ */
+interface TopSitesStorage {
+
+  /**
+   * Return a unified list of top sites based on the given number of sites desired.
+   * If `includeFrecent` is true, fill in any missing top sites with frecent top site results.
+   */
+  fun getTopSites(totalNumberOfSites: Int, includeFrecent: Boolean): List<TopSite>
+
+  interface Observer {
+    /**
+     * Invoked when changes are made to the storage.
+     */
+    fun onStorageUpdated()
+  }
+}
+
+// Already exists in browser-storage-sync.
+interface PlacesHistoryStorage {
+  fun getTopFrecentSites(num: Int)
+}
+
+class DefaultTopSitesStorage(
+  val pinnedSitesStorage: PinnedSitesStorage,
+  val historyStorage: HistoryStorage? = null
```

Comment:
You're the hard-coded default sites. It should also be passed in here (as a `List` or a `Set`), so that the implementation can merge them in when necessary.

Reply:
I agree we need a way to better handle `defaultSites`. For a bit of context, we currently add default sites into the `PinnedSitesStorage` and set a `isDefault` boolean flag to true for these items. See Fenix implementation in https://github.com/mozilla-mobile/fenix/blob/e6a56aec36fba68ebed183c3c268126ece5135df/app/src/main/java/org/mozilla/fenix/components/TopSiteStorage.kt#L56.

One of the requirements would be to allow users to exclude/remove default sites. It is currently easily dealt with by having it inside of `PinnedSitesStorage`, but we will want to re-evaluate that solution with the consideration that default sites can change over time, and we may want to present the updated list of default sites to all users instead of only new installs. Currently, we had one instance where we had changed the default sites in Fenix, but we only cared about new installs so that meant we didn't have to worry about any database migrations and making these changes in A-C.

Reply:
Having the app pass a list of default sites to the feature (or this storage wrapper) solves the migration problem. If that list changes in the app (say, after an update), top sites feature will operate over the new list. So both new installs _and_ existing installs will the new default sites. No db migrations, etc required.